GDB Python API primer

chrisc11 opened the pull request:
Introduce the GDB Python API with two of my personal favorite APIs:

  - pretty printers
  - custom gdb commands

tyhoff commented on _drafts/gdb-python-post.md:
````diff
@@ -0,0 +1,378 @@
+$2 = {next = 0x200070a8, uuid = {bytes = "\235]D@\213Z#^\251\357\354?\221\234zA"}}
+```
+
+We can clean it up a little by enabling pretty printing which will put each field on a newline:
````
Show the command to enable pretty printing in the code snippet below

franc0is replied:
yeah, how do I enable pretty printing?

chrisc11 replied:
I swear that was there at some point hehe, thanks!

tyhoff commented on _drafts/gdb-python-post.md:
````diff
@@ -0,0 +1,378 @@
+       print(self._uuid_list_to_str(node_ptr_val))
+```
+
+To actually load the command all we need to do is instantiate the object by adding:
````
"by adding to the end of the file" or something along those lines.

tyhoff commented on _drafts/gdb-python-post.md:
```diff
@@ -0,0 +1,378 @@
+
+We hope this post gave you a useful overview of how to add custom GDB commands and pretty printers using the Python API. Sometimes adding GDB commands like these can even be used to reduce [code size](https://interrupt.memfault.com/blog/best-firmware-size-tools) by replacing a on-device CLI command that would display the same information!
+
+Do you already have some ideas about how the python API could be applied to automate parts of your debugging flow ... perhaps a command to walk custom heaps or display the contents in a memory-mapped filesystem? Or maybe you are already have some great examples of how you have used GDB Python? Either way, let us know in the discussion area below!
```
Make sure to add a link to an example repo with the source code and the exactly Python script they could source.

franc0is replied:
You could add the source files to the `interrupt` repo since it's open source already

chrisc11 replied:
Works for me! I was going to tack it into zero-to-main but I thing it probably does just make more sense to have an example/ directory in this repo

franc0is commented on _drafts/gdb-python-post.md:
```diff
@@ -0,0 +1,378 @@
+
+While all this manual typing may be fine for debugging one crash or issue, it does not scale. When a new firmware is being QA'd for the first time on hundreds of devices internally, it's important to be able to quickly classify if the issue is known or not so debug doesn't cause the schedule to slip. Even once a product makes it into the wild, an issue that appears on .1% of devices in a one mill ...
+
+Manually inspecting hundreds of crashes is no fun ... but fortunately, GDB has several facilities built in that allow you to [extend](https://sourceware.org/gdb/onlinedocs/gdb/Extending-GDB.html#Extending-GDB) its core functionality, including Python, the scripting language of choice for many embedded projects. This allows one to accelerate and in many cases even **automate** debug!
```
This sentence is a bit odd:
> GDB has several facilities built in that allow you to extend its core functionality, including Python, the scripting language of choice for many embedded project.

I suggest:
> GDB allows user to extend its core functionality in a number of ways, including via its Python API.

franc0is commented on _drafts/gdb-python-post.md:
```diff
@@ -0,0 +1,378 @@
+
+Let's walk through a trivial example snippet of code to get an idea of how some of the APIs work!
+
+The snippet tracks UUIDs in RAM in a linked list. On a real embedded system, we may have a bunch of items, identified by their UUID, stored on an external flash. If the list of items are looked at frequently, it may make sense for the list may be cached in RAM for faster lookup (since flash access can be slow) and better power (since reading from external flash consumes more power than a RAM read ...
```
Another weird sentence:
> if the list of items **are** looked at frequently, it **may make sense for the list** may be cached in RAM for faster lookup [...]

Should be
> If the list of items **is** looked at frequently, it may be cached in RAM for faster access and better power.

franc0is commented on _drafts/gdb-python-post.md:
```diff
@@ -0,0 +1,378 @@
+
+void list_add_uuid(const Uuid *fake_uuid) {
+  UuidListNode *node_to_add = malloc(sizeof(UuidListNode));
+  if (node_to_add == NULL) {
```
maybe assert?

franc0is commented on _drafts/gdb-python-post.md:
```diff
@@ -0,0 +1,378 @@
+
+static UuidListNode *s_list_head = NULL;
+
+void list_add_uuid(const Uuid *fake_uuid) {
```
I wouldn't call this argument `fake_uuid` since this is a perfectly fine "list add" function.

franc0is commented on _drafts/gdb-python-post.md:
```diff
@@ -0,0 +1,378 @@
+
+### GDB Pretty Printers
+
+When we print a structure, GDB will by default will display each of the individual fields. The `s_list_head` in the code above would look something like this:
```
> will by default will display [...]

change to

> GDB will display each of the individual fields

chrisc11 commented:
@franc0is @tyhoff I think I addressed all the code review commits if you want to give it a final pass (top commit)

I also looked at the jekyll output and reflowed some of the code blocks so they would fit w/o needing to scroll horizontally

tyhoff commented on _drafts/gdb-python-post.md:
```diff
@@ -0,0 +1,389 @@
+
+While all this manual typing may be fine for debugging one crash or issue, it does not scale. When a new firmware is being QA'd for the first time on hundreds of devices, it's important to be able to quickly classify the issue so time spent debugging doesn't cause the schedule to slip. Even once a product makes it into the wild, an issue that appears on .1% of devices in a one million device flee ...
+
+Manually inspecting hundreds of crashes is no fun ... but fortunately, GDB allows users to extend its core functionality in a number of ways, including via its [Python API](https://sourceware.org/gdb/onlinedocs/gdb/Extending-GDB.html#Extending-GDB) . This allows one to accelerate and in many cases even **automate** debug!
```
**automate** debugging

franc0is commented:
Shipit!
